creator-node/src/URSMService.js

```diff
@@ -0,0 +1,234 @@
+ *
+ * Steps:
+ * - If already registered on URSM, exit successfully
+ * - Submit requests for proposal to 3 nodes already registered on URSM
+ *    - keep attempting to submit requests until 3 successful or no remaining nodes
+ * - Register self on URSM with 3 proposer signatures
+ */
+class URSMService {
+  constructor (nodeConfig, audiusLibs) {
+    this.nodeConfig = nodeConfig
+    this.audiusLibs = audiusLibs
+    this.spID = nodeConfig.get('spID')
+    this.delegateOwnerWallet = nodeConfig.get('delegateOwnerWallet')
+    this.delegatePrivateKey = nodeConfig.get('delegatePrivateKey')
+    this.spOwnerWallet = nodeConfig.get('spOwnerWallet')
+
+    // Requires properly initialized libs instance and non-zero spID
+    // Currently spID is configured inside snapbackSM
+    if (!this.audiusLibs || !this.spID || !this.delegateOwnerWallet || !this.delegatePrivateKey || !this.spOwnerWallet) {
+      throw new Error('nope')
+    }
+
+    this.numSignaturesRequired = 3
+
+    this.logInfo(`URSMSERVICE CONSTRUCTED`)
+  }
+
+  logInfo (msg) {
+    logger.info(`URSMService || ${msg}`)
+  }
+
+  logError (msg) {
+    logger.error(`URSMService ERROR || ${msg}`)
+  }
+
+  /**
+   * TODO DESCRIPTION
+   *
+   * Steps: 
+   *  1. Fetch node record from L1 ServiceProviderFactory for spID
+   *    a. Short-circuit if no L1 record found
+   *  2. Fetch node record from L2 UserReplicaSetManager for spID
+   *    a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
+   *  3. Fetch list of all nodes registered on URSM, in order to submit requests for proposal
+   *    a. Randomize list to minimize bias
+   *    b. Organize list to ensure requests are sent to nodes with unique ownerWallets
+   *  4. Submit requests for proposal to nodes until 3 successful signatures received
+   *    a. Must have signatures from nodes with unique ownerWallets
+   *    b. Error if all available nodes contacted without 3 successful signatures
+   *  5. Submit registration transaction to URSM with signatures
+   */
+  async init () {
+    this.logInfo('STARTING INIT')
+
+    /**
+     * 1. Fetch node record from L1 ServiceProviderFactory for spID
+     */
+    const spRecordFromSPFactory = await this.audiusLibs.ethContracts.ServiceProviderFactoryClient.getServiceEndpointInfo(
+      'content-node',
+      this.spID
+    )
+    let {
+      owner: ownerWalletFromSPFactory,
+      delegateOwnerWallet: delegateOwnerWalletFromSPFactory,
+      endpoint: endpointFromSPFactory
+    } = spRecordFromSPFactory
+    delegateOwnerWalletFromSPFactory = delegateOwnerWalletFromSPFactory.toLowerCase()
+
+
+    /**
+     * 1-a. Short-circuit if no L1 record found
+     */
+    if (
+      LibsUtils.isZeroAddress(ownerWalletFromSPFactory)
+      || LibsUtils.isZeroAddress(delegateOwnerWalletFromSPFactory)
+      || !endpointFromSPFactory
+    ) {
+      throw new Error('L1recordbad')
```

Comment:
probably want to validate that the endpointFromSPFactory is actually a match to the node's configured endpoint too - remember that the spID config is always recalculated from the endpoint and is not provided by the user